File: deck/base/code/native/node/runtime/path.ts
// Path manipulation over node:path, presented as one namespace so the native module docks it as a <global:path> and
// forwards uniformly, matching the rust, swift, and kotlin path shims. The file extension carries its leading dot.
import {
  join as joinPaths,
  dirname,
  basename,
  extname,
  isAbsolute,
} from 'node:path'

const path = {
  join: (base: string, name: string): string => joinPaths(base, name),
  directory: (target: string): string => dirname(target),
  fileName: (target: string): string => basename(target),
  fileExtension: (target: string): string => extname(target),
  isAbsolute: (target: string): boolean => isAbsolute(target),
}

File: deck/base/code/native/swift/runtime/path.swift
import Foundation

// Path manipulation over Foundation's NSString path APIs. The file extension carries its leading dot.
enum path {
    static func join(_ base: String, _ name: String) -> String {
        return (base as NSString).appendingPathComponent(name)
    }
    static func directory(_ target: String) -> String {
        return (target as NSString).deletingLastPathComponent
    }
    static func fileName(_ target: String) -> String {
        return (target as NSString).lastPathComponent
    }
    static func fileExtension(_ target: String) -> String {
        let value = (target as NSString).pathExtension
        return value.isEmpty ? "" : "." + value
    }
    static func isAbsolute(_ target: String) -> Bool {
        return (target as NSString).isAbsolutePath
    }
}
